### PR TITLE
osx support (fixes #128)

### LINKED ISSUE #128: OSX support
Get Envoy compiling on OSX. Shim out all linux specific features and either turn them off in OSX build or provide OS shim layer to make them work on BSD.

### PR DESCRIPTION
**Pull request by zuercher**
Some highlights:

source/common/filesystem: adds a kqueue-based implementation of WatcherImpl. The conditional compiliation is hacky.

source/common/network: handle odd-ball FreeBSD/osx errors on get/setsockopt. Also, SOCK_NONBLOCK doesn't exist on FreeBSD or OSX. OSX has an extra byte field in `struct sockaddr` and derivatives that needs to be accounted for in some places.

source/common/runtime: use libuuid to generate UUIDs on Linux and OSX. Freebsd gets a hand-rolled implementation.

source/exe: hot restart is completely disabled for osx and freebsd

test/common/http: on OSX, `struct tm` configured for a date before 1901-12-13 20:45:52 becomes a `time_t` that represents 1969-12-31 23:59:59.

test/common/network: OSX doesn't like some of IPv6 addresses used in testing. By default only a single IPv4 loopback address exists in OSX (127.0.0.1).

Various and sundry explicit #include statements to pick up definitions that are obtained transitively on Linux.

### CODE REVIEW

File: patches/freebsd11_osymlink.patch [hunk 1]
```diff
@@ -0,0 +1,95 @@
+--- ./lib/libc/sys/open.2.orig	2016-09-28 16:25:57.000000000 -0700
```
Comment (htuch) [inline patches/freebsd11_osymlink.patch]: What is this file intended for?

File: source/common/filesystem/watcher_impl.cc [hunk 2]
```diff
@@ -1,93 +1,7 @@
-#include "common/filesystem/watcher_impl.h"
+// NOLINT(namespace-envoy)
+#if defined(__linux__)
+#include "common/filesystem/watcher_impl_linux.cc"
```
Comment (htuch) [inline source/common/filesystem/watcher_impl.cc]: Can we use select based Bazel target srcs/hdrs to avoid this kind of compile time conditioning?

File: source/common/network/address_impl.cc [hunk 3]
```diff
@@ -78,7 +82,14 @@ InstanceConstSharedPtr peerAddressFromFd(int fd) {
 }
 
 int InstanceBase::flagsFromSocketType(SocketType type) const {
+#if defined(__FreeBSD__)
+  int flags = O_NONBLOCK;
+#elif defined(__APPLE__)
```
Comment (htuch) [inline source/common/network/address_impl.cc]: In various places you condition on both __FreeBSD__ and __APPLE__. Why is this? Are you adding FreeBSD and OS X support? Is the FreeBSD part of the OS X support (but then why special __APPLE__ conditions?)?

File: source/common/network/address_impl.cc [hunk 4]
```diff
@@ -130,7 +141,13 @@ int Ipv4Instance::connect(int fd) const {
 }
 
 int Ipv4Instance::socket(SocketType type) const {
-  return ::socket(AF_INET, flagsFromSocketType(type), 0);
+  int fd = ::socket(AF_INET, flagsFromSocketType(type), 0);
+#ifdef __APPLE__
+  if (fd != -1) {
+    RELEASE_ASSERT(fcntl(fd, F_SETFL, O_NONBLOCK) != -1);
```
Comment (htuch) [inline source/common/network/address_impl.cc]: Can you add a comment explaining why this is a special case for __APPLE__?

File: source/common/network/connection_impl.cc [hunk 5]
```diff
@@ -438,6 +450,27 @@ void ConnectionImpl::onWriteReady() {
     ASSERT(0 == rc);
     UNREFERENCED_PARAMETER(rc);
 
+#if !defined(__linux__)
+    // When a connect() is performed on a non-blocking socket, it'll return -1/EINPROGRESS
+    // immediately and when select()'ed for writability it'll succeed or fail but when it fails the
+    // reason is hidden. `getsockopt(,,SO_ERROR,,)` solves this issue on some systems but on some
+    // cases such as FreeBSD getsockopt returns 0 even on `EAGAIN`. The following trick issues a
+    // `getpeername()` which will return -1/ENOTCONN if the socket is not connected and read(,,1)
+    // will produce the right errno.
```
Comment (htuch) [inline source/common/network/connection_impl.cc]: This seems sketchy. Is there some reference to where this is shown to be accurate for all errnos?

File: source/common/network/connection_impl.cc [hunk 6]
```diff
@@ -438,6 +450,27 @@ void ConnectionImpl::onWriteReady() {
     ASSERT(0 == rc);
     UNREFERENCED_PARAMETER(rc);
 
+#if !defined(__linux__)
```
Comment (htuch) [inline source/common/network/connection_impl.cc]: Is this true for all non-Linux or only for OS X?

File: source/common/runtime/runtime_impl.cc [hunk 7]
```diff
@@ -22,20 +26,19 @@ namespace Runtime {
...
-    throw EnvoyException(fmt::format("cannot read the uuid: bytes read - {}, bytes expected - {}",
-                                     bytes_read, UUID_LENGTH));
-  }
+#if defined(__linux__) || defined(__APPLE__)
+  uuid_t uuid;
+  uuid_generate_random(uuid);
+  uuid_unparse(uuid, generated_uuid);
+#else
+  sprintf(generated_uuid, "%08lx-%04x-%04x-%04x-%04x%08lx",
+          static_cast<unsigned long>(arc4random()), static_cast<unsigned>(arc4random() & 0xffff),
```
Comment (htuch) [inline source/common/runtime/runtime_impl.cc]: Probably best to generate the minimum number of random numbers required here before the `sprintf`. Also, would prefer `snprintf` if using C formatters for safety.

File: source/exe/hot_restart.h [hunk 8]
```diff
@@ -201,5 +205,39 @@ class HotRestartImpl : public HotRestart,
   bool parent_terminated_{};
 };
 
+#else  // __linux__
+
+/**
+ * No-op implementation of HotRestart for everybody else.
```
Comment (htuch) [inline source/exe/hot_restart.h]: We (@ Google) would find this useful even on Linux. @mrice32. Maybe there is some way to condition this at build time.

File: source/server/options_impl.cc [hunk 9]
```diff
@@ -11,6 +11,10 @@
 #include "spdlog/spdlog.h"
 #include "tclap/CmdLine.h"
 
+#if defined(__APPLE__)
+#define uint64_t unsigned long
```
Comment (lizan) [inline source/server/options_impl.cc]: Why do you need this? `uint64_t` should just exists.

File: source/exe/signal_action.cc [hunk 10]
```diff
@@ -16,6 +16,8 @@ void SignalAction::sigHandler(int sig, siginfo_t* info, void* context) {
 #ifdef REG_RIP
     // x86_64
     error_pc = reinterpret_cast<void*>(ucontext->uc_mcontext.gregs[REG_RIP]);
+#elif defined(__APPLE__)
```
Comment (dnoe) [inline source/exe/signal_action.cc]: Do you want to add a conditional for x86 32 bit here? No need to implement it now, but I'm guessing a 32 bit compile will fail because there's no __rip.

File: source/exe/signal_action.cc [hunk 11]
```diff
@@ -57,13 +59,24 @@ void SignalAction::installSigHandlers() {
...
   RELEASE_ASSERT(sigaltstack(&previous_altstack_, nullptr) == 0);
+
   int hidx = 0;
   for (const auto& sig : FATAL_SIGS) {
     RELEASE_ASSERT(sigaction(sig, &previous_handlers_[hidx++], nullptr) == 0);
   }
 }
 
+#ifdef __APPLE__
+#define MAP_STACK (0)
```
Comment (dnoe) [inline source/exe/signal_action.cc]: Does it make sense to add #ifndef MAP_STACK around this in case Darwin adds real map_stack in the future?

File: source/exe/signal_action.cc [hunk 12]
```diff
@@ -57,13 +59,24 @@ void SignalAction::installSigHandlers() {
 }
 
 void SignalAction::removeSigHandlers() {
+#if defined(__APPLE__)
+  // ss_flags contains SS_DISABLE, but Darwin still checks the size, contrary to the man page
```
Comment (dnoe) [inline source/exe/signal_action.cc]: D'oh, good catch!

File: source/exe/signal_action.h [hunk 13]
```diff
@@ -47,7 +49,9 @@ namespace Envoy {
 class SignalAction : NonCopyable {
 public:
   SignalAction()
-      : guard_size_(sysconf(_SC_PAGE_SIZE)), altstack_size_(guard_size_ * 4), altstack_(nullptr) {
+      : guard_size_(sysconf(_SC_PAGE_SIZE)),
+        altstack_size_(std::max(guard_size_ * 4, static_cast<size_t>(MINSIGSTKSZ))),
```
Comment (dnoe) [inline source/exe/signal_action.h]: I think if the MINSIGSTKSZ case is hit then protections for the immediately adjacent pages still works OK, there's just a portion at the end which is not protected. The second mprotect call above could have the size adjusted to cover this case.

File: source/server/options_impl.cc (hunk 9 above)
Comment (dnoe) [inline source/server/options_impl.cc]: This should be in header `<cstdint>` included above. You might need to do `using std::uint64_t;` however.

File: test/integration/hotrestart_test.sh [hunk 14]
```diff
@@ -2,6 +2,12 @@
 
 set -e
 
+if [[ `uname` != "Linux" ]]
```
Comment (dnoe) [inline test/integration/hotrestart_test.sh]: I think there's probably a way this test could be completely disabled with a Bazel rule instead.

@htuch

File: source/common/mongo/bson_impl.cc [hunk 15]
```diff
@@ -10,6 +10,15 @@
 
 #include "spdlog/spdlog.h"
 
+#ifdef __APPLE__
```
Comment (mattklein123) [inline source/common/mongo/bson_impl.cc]: This will end up getting used again elsewhere. Can we put this into common/common/byte_order.h or something?

File: source/common/network/connection_impl.cc [hunk 16]
```diff
@@ -177,6 +177,18 @@ void ConnectionImpl::noDelay(bool enable) {
   // Set NODELAY
   int new_value = enable;
   rc = setsockopt(fd_, IPPROTO_TCP, TCP_NODELAY, &new_value, sizeof(new_value));
+  // Since we are using SOCK_NONBLOCK, there are cases (on FreeBSD for example) where the immediate
+  // connect response is EINPROGRESS but the socket failed and the calling code might not know about
+  // it yet.
+  if (-1 == rc && errno == ECONNRESET) {
```
Comment (mattklein123) [inline source/common/network/connection_impl.cc]: Can we guard this with free bsd only then?

File: source/common/network/address_impl.cc [hunk 17]
```diff
@@ -130,7 +141,13 @@ int Ipv4Instance::connect(int fd) const {
 }
 
 int Ipv4Instance::socket(SocketType type) const {
-  return ::socket(AF_INET, flagsFromSocketType(type), 0);
+  int fd = ::socket(AF_INET, flagsFromSocketType(type), 0);
+#ifdef __APPLE__
+  if (fd != -1) {
```
Comment (mattklein123) [inline source/common/network/address_impl.cc]: is there any point in doing fd != -1 check here? (Same in other places). Will just crash anyway in RELEASE_ASSERT.

File: source/common/network/utility.cc [hunk 18]
```diff
@@ -268,13 +272,26 @@ Address::InstanceConstSharedPtr Utility::getAddressWithPort(const Address::Insta
...
-    // TODO(mattklein123): IPv6 support. See github issue #1094.
-    ASSERT(orig_addr.ss_family == AF_INET);
-    return Address::InstanceConstSharedPtr{
-        new Address::Ipv4Instance(reinterpret_cast<sockaddr_in*>(&orig_addr))};
+    switch (orig_addr.ss_family) {
+    case AF_INET:
+      return Address::InstanceConstSharedPtr{
+          new Address::Ipv4Instance(reinterpret_cast<sockaddr_in*>(&orig_addr))};
+
+    case AF_INET6:
```
Comment (mattklein123) [inline source/common/network/utility.cc]: Do these new cases and the throw have code coverage on linux CI? If not will need to refactor somehow to get coverage.

File: source/exe/hot_restart.h (hunk 8 above)
Comment (mattklein123) [inline source/exe/hot_restart.h]: On this point, can we potentially just revert all the changes to hot_restart_impl.h/cc, and just add a new hot_restart_nop_impl.h/cc, and do select() compile? This will keep existing code cleaner and make it possible to compile out hot restart support. (I think MSFT people will want this too initially).

File: test/common/network/cidr_range_test.cc [hunk 19]
```diff
@@ -323,14 +323,14 @@ TEST(Ipv6CidrRange, InstanceConstSharedPtrAndLengthCtor) {
 
 TEST(Ipv6CidrRange, StringAndLengthCtor) {
   CidrRange rng;
-  rng = CidrRange::create("::ffff", 122); // Assignment operator.
+  rng = CidrRange::create("ff::ffff", 122); // Assignment operator.
```
Comment (mattklein123) [inline test/common/network/cidr_range_test.cc]: Are we losing any linux test coverage w/ these changes? cc @hennna

File: test/common/upstream/ring_hash_lb_test.cc [hunk 20]
```diff
@@ -102,6 +103,48 @@ TEST_F(RingHashLoadBalancerTest, Basic) {
     TestLoadBalancerContext context(0);
     EXPECT_EQ(cluster_.hosts_[5], lb_.chooseHost(&context));
   }
+#else
```
Comment (mattklein123) [inline test/common/upstream/ring_hash_lb_test.cc]: Not sure if it's really worth doing this. This is already not great and related to my TODO above. Seems like it might be better to just not run these tests on Apple until someone fixes the TODO to compile murmur3 or city into the code (which is not difficult).

File: patches/freebsd11_osymlink.patch (hunk 1 above)
Comment (zuercher) [inline patches/freebsd11_osymlink.patch]: It was in @Reflejo's original branch. I gather it's a FreeBSD path required for the kqueue WatcherImpl to work. I'm fine to remove it.

Comment (Reflejo) [inline patches/freebsd11_osymlink.patch]: It's a patch for freebsd11's kernel to enable `O_SYMLINK`

File: source/common/network/address_impl.cc (hunk 3 above)
Comment (Reflejo) [inline source/common/network/address_impl.cc]: I think we should keep it since I got freeBSD to pass all tests as well. I'd double check that it's still the case. The delta between FreeBSD support and OSX is actually pretty small

File: source/common/network/connection_impl.cc (hunk 5 above)
Comment (Reflejo) [inline source/common/network/connection_impl.cc]: There is a good summary here: https://cr.yp.to/docs/connect.html

File: test/common/network/cidr_range_test.cc (hunk 19 above)
Comment (hennna) [inline test/common/network/cidr_range_test.cc]: Not off the top of my head. I don't think tests are dependent on the address value. Is the issue with OSX the leading ::?

File: source/common/network/connection_impl.cc (hunk 5 above)
Comment (PiotrSikora) [inline source/common/network/connection_impl.cc]: I believe that connection error will be available in `kq_errno` on OSX & FreeBSD and should be used instead of `getsockopt(SO_ERROR)` on those platforms, so this might not be necessary at all.

File: patches/freebsd11_osymlink.patch (hunk 1 above)
Comment (PiotrSikora) [inline patches/freebsd11_osymlink.patch]: Can't we watch the directory (similarly to the inotify implementation on Linux)? I don't think patching FreeBSD kernel and adding custom features to it is a viable prerequisite for Envoy on FreeBSD.

Comment (Reflejo) [inline patches/freebsd11_osymlink.patch]: Yea I didn't intended this to go to upstream. Honestly I don't think there is a way to do what envoy is doing for the symlink observation without this. But I didn't investigate this longer enough for a conclusive answer.

Comment (PiotrSikora) [inline patches/freebsd11_osymlink.patch]: To spin this differently... @mattklein123: now, that we have LDS and overall better xDS support, do you think that we still need the runtime configuration (and therfore `Watcher`)?

Comment (mattklein123) [inline patches/freebsd11_osymlink.patch]: Yes, we need it. We will also support loading xDS APIs off of filesystem. Either way, I think we can ignore BSD for the purpose of this PR. Let's just get OSX working and move in. If someone like @Reflejo wants to add in BSD support after and can test on current master, that's fine.

File: source/common/network/address_impl.cc (hunk 3 above)
Comment (zuercher) [inline source/common/network/address_impl.cc]: going to remove the partial FreeBSD support per gitter conversation

File: source/common/network/connection_impl.cc (hunk 6 above)
Comment (zuercher) [inline source/common/network/connection_impl.cc]: I believe it's true for FreeBSD and OS X. I'll just change it to #ifdef \_\_APPLE\_\_.

File: source/common/runtime/runtime_impl.cc (hunk 7 above)
Comment (zuercher) [inline source/common/runtime/runtime_impl.cc]: I'll just remove this since it's not used for Linux or OS X.

File: source/common/network/address_impl.cc (hunk 4 above)
Comment (zuercher) [inline source/common/network/address_impl.cc]: Will do.

File: source/common/network/connection_impl.cc (hunk 5 above)
Comment (zuercher) [inline source/common/network/connection_impl.cc]: I can only offer that it works, empirically. This case is triggered regularly in testing on OS X. I think it makes sense to keep it disabled under Linux since it'll only add overhead and doesn't seem to be a problem there.

File: source/common/network/address_impl.cc (hunk 17 above)
Comment (zuercher) [inline source/common/network/address_impl.cc]: I did the check because the original code didn't assert (for AF_INET and AF_UNIX). I'll go ahead and just assert on the fd and then assert on the fcntl call.

File: source/server/options_impl.cc (hunk 9 above)
Comment (zuercher) [inline source/server/options_impl.cc]: `uint64_t` exists but is typedef `unsigned long long`. Without this you get this error:

```
In file included from source/server/options_impl.cc:12:
In file included from external/envoy_deps/thirdparty/tclap/include/tclap/CmdLine.h:27:
In file included from external/envoy_deps/thirdparty/tclap/include/tclap/SwitchArg.h:30:
In file included from external/envoy_deps/thirdparty/tclap/include/tclap/Arg.h:54:
external/envoy_deps/thirdparty/tclap/include/tclap/ArgTraits.h:80:22: error: type 'unsigned long long' cannot be used prior to '::' because it has no members
    typedef typename T::ValueCategory ValueCategory;
                     ^
external/envoy_deps/thirdparty/tclap/include/tclap/ValueArg.h:403:37: note: in instantiation of template class 'TCLAP::ArgTraits<unsigned long long>' requested here
        ExtractValue(_value, val, typename ArgTraits<T>::ValueCategory());
                                           ^
external/envoy_deps/thirdparty/tclap/include/tclap/ValueArg.h:363:5: note: in instantiation of member function 'TCLAP::ValueArg<unsigned long long>::_extractValue' requested here
                                _extractValue( args[*i] );
                                ^
source/server/options_impl.cc:31:29: note: in instantiation of member function 'TCLAP::ValueArg<unsigned long long>::processArg' requested here
  TCLAP::ValueArg<uint64_t> base_id(
                            ^
1 error generated.
```

Swapping out the typedef I added for `using std::uint64_t;` produces an identical error.

File: source/common/network/connection_impl.cc (hunk 16 above)
Comment (zuercher) [inline source/common/network/connection_impl.cc]: I've removed it since it doesn't seem to happen on OSX.

File: test/integration/hotrestart_test.sh (hunk 14 above)
Comment (zuercher) [inline test/integration/hotrestart_test.sh]: Yes, but I don't know how.

Normally, I could use `select` to choose either no test script or a different one. However, the output of `select` is only usable by bazel rules (because only at the point of rule execution is it possible to inspect the configuration to choose which subset of parameters to select).

`envoy_sh_test` is a macro and it attempts to access members of its `srcs` parameter to produce inputs for several rules. This doesn't work if srcs is assigned from select (whose output has no `[]` operator). To make it work, envoy_sh_test would have to become a rule and at that point I'm in over my head.

File: source/exe/signal_action.h (hunk 13 above)
Comment (zuercher) [inline source/exe/signal_action.h]: I think it's correct. We ultimately mmap `altstack_size_ + 2 * guard_size_` bytes (see `mapSizeWithGuards`) so `altstack_size_` doesn't include the guard pages. The second mprotect calls protects `guard_size_` bytes at an address that's `guard_size_ + altstack_size_` bytes from the start of the mapped memory.

We call `sigaltstack` with `altstack_ + guard_size_` (to skip the first guard page) and `altstack_size_` (which doesn't count the guard pages, and so stops just before the second guard page).

Did I miss something?

File: source/exe/signal_action.cc (hunk 11 above)
Comment (zuercher) [inline source/exe/signal_action.cc]: Will do.

File: source/exe/signal_action.cc (hunk 10 above)
Comment (zuercher) [inline source/exe/signal_action.cc]: Sure.

File: source/exe/hot_restart.h [hunk 21]
```diff
@@ -201,5 +122,39 @@ class HotRestartImpl : public HotRestart,
...
+  void terminateParent() override {}
+  void shutdown() override {}
+  std::string version() override { return "disabled"; }
+
+  // RawStatDataAllocator
+  Stats::RawStatData* alloc(const std::string& name) override;
+  void free(Stats::RawStatData& data) override;
+
+private:
+  SharedMemory& shmem_;
```
Comment (mattklein123) [inline source/exe/hot_restart.h]: IMO the NOP hot restart impl should not use shared memory or the process shared mutex stuff. Basically, I would just do exactly what https://github.com/lyft/envoy/blob/master/test/integration/server.cc#L93 does which is to use process local locks, the heap allocator, etc. I think you can probably revert all the changes related to shared memory and just move the test impl into prod code somewhere and use that.

**Comment by mattklein123** [comment]
@zuercher general comment. Can we start to split out some of this change into self contained PRs? Some of the test fixes can go in and get reviewed independently, as can some of the trivial header changes, etc. That way we can start merging the not controversial stuff and trim down the size of the main PR. Thank you.

File: test/common/network/cidr_range_test.cc (hunk 19 above)
Comment (zuercher) [inline test/common/network/cidr_range_test.cc]: Sort of. [Wikipedia](https://en.wikipedia.org/wiki/IPv6_address#Default_address_selection) tells me an old version of the IPv6 spec considered `::/96` addresses (that is addresses with 96 bits of leading zeros) to be IPv6-mapped IPv4 addresses. It's been deprecated for a decade, but when OSX converts these addresses to a string they comes out as "::0.0.255.255" and the tests fail because they do string comparisons. Linux only provides the "mapped IPv4" style addresses with trailing dotted quads for the `::ffff:0:0/96` range (which replaced `::/96`). It seems to me that these test are concerned with how the low order bits of the address are manipulated and that using a different prefix doesn't change the meaning of the tests.

(I wrote a little program to round trip IP address strings to binary and back. https://gist.github.com/zuercher/0d0bc5a433fda423124c71afa2eb29e1)

File: source/exe/signal_action.h (hunk 13 above)
Comment (dnoe) [inline source/exe/signal_action.h]: Ah, yes, all of this looks correct. Guard pages are always outside the altstack_size_, and MINSIGSTKSZ should always be a multiple of page size.

File: source/server/options_impl.cc (hunk 9 above)
Comment (dnoe) [inline source/server/options_impl.cc]: Hrrm, this is unfortunate. I don't understand why the define doesn't cause the same problem.

If we can't figure out a better solution please do add a detailed comment explaining why it's done this way.

**Comment by zuercher** [comment]
Split out #1363 and #1365 for the watcher and hot restart changes. (I'll split more things out tomorrow.)

File: source/server/options_impl.cc (hunk 9 above)
Comment (zuercher) [inline source/server/options_impl.cc]: Finally figured this out. TCLAP provides implementations of `ArgTraits<T>` for various types, but the ones for `unsigned long long` and `long long` are gated by `#ifdef HAVE_LONG_LONG`, which isn't set. I think they were expecting users of libraries to use autoconf which would detect this and make the necessary #define. I'll post it with a few other things tomorrow (split out in a new PR).

**Comment by mattklein123** [comment]
Going to close this out in lieu of the other PRs that are ongoing.
